### PR TITLE
Push current progress for inspection

### PR DESCRIPTION
**Pull request by sophie-xhonneux**
## Issue Number

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/trainer.py [hunk 1]
```diff
@@ -375,7 +375,16 @@ def train(self, mini_epoch):
                     batch, self.model_params, self.model
                 )
 
-            loss = self.loss_calculator.compute_loss(preds=preds, targets=targets_and_auxs)
+            loss = self.loss_calculator.compute_loss(
+                preds=preds,
+                targets=targets_and_auxs,
+                metadata=(
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Can't we extract the metadata in the target_aux_calculator as aux information? That seems more natural and we anyway need to touch it there.

Comment (clessig) [reply src/weathergen/train/trainer.py]: Otherwise, we should at least put it into a function

File: src/weathergen/train/loss_calculator.py [hunk 2]
```diff
@@ -56,38 +56,29 @@ def __init__(
...
-        ]
+        calculator_configs = (
+            cf.training_config.losses if stage == TRAIN else cf.validation_config.losses
+        )
 
         self.loss_calculators = [
-            (config.weight, Cls(cf=cf, loss_fcts=config.loss_fcts, stage=stage, device=self.device))
-            for (Cls, config) in calculator_configs
+            (
+                config.pop("weight"),
```
Comment (clessig) [inline src/weathergen/train/loss_calculator.py:65]: That's not elegant. Can't we avoid this?

File: src/weathergen/train/loss_modules/__init__.py [hunk 3]
```diff
@@ -7,6 +7,27 @@
 # granted to it by virtue of its status as an intergovernmental organisation
 # nor does it submit to any jurisdiction.
 
+from .loss_module_base import LossModuleBase, LossValues
 from .loss_module_physical import LossPhysical
 
-__all__ = [LossPhysical]
+
+class LossLatentSSLStudentTeacher(LossModuleBase):
```
Comment (clessig) [inline src/weathergen/train/loss_modules/__init__.py]: This should come in the student-teacher PR

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 4]
```diff
@@ -571,22 +632,22 @@ def _get_batch(self, idx: int, forecast_dt: int):
...
 
             # tokenize windows
             # *_tokens = [ (cells_idx, cells_idx_lens), ... ] with length = #time_steps
             input_tokens = self.tokenizer.get_tokens_windows(stream_info, input_data, True)
             output_tokens = self.tokenizer.get_tokens_windows(stream_info, output_data, False)
 
-            # collect source data for current stream
-            # loop over student views
             for sidx, source_mask in enumerate(source_masks.masks):
+                # Map each student (source) to its teacher (target)
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: No student/teacher here: source and targets are the terms to be used.

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 5]
```diff
@@ -571,22 +632,22 @@ def _get_batch(self, idx: int, forecast_dt: int):
...
+                tidx = student_to_teacher[sidx].item()
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: Same here, this should not be called `student_teacher`

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 6]
```diff
@@ -597,42 +658,42 @@ def _get_batch(self, idx: int, forecast_dt: int):
...
-                # also want to add the mask to the metadata
-                source_metadata = source_masks.metadata[sidx]
-                source_metadata.mask = source_mask
+                batch.add_source_stream(sidx, tidx, stream_name, sdata, source_masks.metadata[sidx])
 
-                # map each source to its target
-                t_idx = student_to_teacher[sidx]
-                batch.add_source_stream(sidx, t_idx, stream_name, sdata, source_metadata)
+            # stream_data_target can contain network input
+            stream_data_target = {}
```
Comment (clessig) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: Remove. This doesn't exist any longer; it's all in batch

File: src/weathergen/train/loss_modules/__init__.py (hunk 3 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/loss_modules/__init__.py]: I need a mock-up to test the configs, but I can remove

File: src/weathergen/train/loss_calculator.py (hunk 2 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/loss_calculator.py:65]: I didn't find a good way, for when you have multiple Loss calculators

File: src/weathergen/train/loss_modules/__init__.py (hunk 3 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/loss_modules/__init__.py]: would you like me to remove the mock-up?
